File: Week 02/id_118/Python/LeetCode_46_118.py
from typing import List


class Solution:
    """
    This solution follows strictly the generic recursion solution from the class
    Time Complexity: O(n^n) exponential
    Space Complexity: O(n)
    """

    def __init__(self):
        # initialize result in res, as a list of list
        self.res = []
        # initialize a helper vector
        self.used = []

    def _gen_perm(self, curr_level, max_level, nums, curr_p):
        # terminator
        if curr_level == max_level:
            # this is important to add a copy method,
            # otherwise it will be affected by the pop later
            # Assignment statements in Python do not copy objects,
            # they create bindings between a target and an object.
            self.res.append(curr_p.copy())
            return
        # process the current level
        for i in range(len(nums)):
            if not self.used[i]:
                # if nums[i] is not used
                curr_p.append(nums[i])
                self.used[i] = True

                # drill down
                self._gen_perm(curr_level + 1, max_level, nums, curr_p)

                # really important to reverse and reset global variables here
                curr_p.pop()
                self.used[i] = False
        return

    def permute(self, nums: List[int]) -> List[List[int]]:
        if len(nums) == 0:
            return self.res

        self.used = [False] * len(nums)
        # current solution is []
        self._gen_perm(0, len(nums), nums, [])

        return self.res


if __name__ == '__main__':
    sol = Solution()
    print(sol.permute([1, 2, 3]))

File: Week 02/id_118/C++/LeetCode_46_118.cpp
//
// Created by Isaac Zhou on 2019-10-26.
//

#include <iostream>
#include <vector>
#include <cassert>
#include <unordered_map>

using namespace std;

class Solution {

private:
    // create a vector of vectors to store result
    vector <vector<int>> res;
    // create a helper vector to see if number is already used in permutation
    vector<bool> used;

    // Recursion method
    void genPerm(const <int> &nums, int index, vector<int> &p) {
        // terminator
        if (index == nums.size()) {
            res.push_back(p);
            return;
        }
        for (int i = 0; i < nums.size(); ++i) {
            if (!used[i]) {
                // process the current logic
                // add nums[i] to p only if nums[i] is not used yet
                p.push_back(nums[i]);
                // update used[i] status
                used[i] = true;

                // drill down
                genPerm(nums, index + 1, p);

                // reverse, because we changed the p and used status
                // reset them
                p.pop_back();
                used[i] = false;
            }
        }
    }

public:
    vector <vector<int>> permute(vector<int> &nums) {
        // initialize res with empty
        res.clear();
        // edge case
        if (nums.size() == 0) {
            return res;
        }
        // create a vector p to store solution for each recursion
        vector<int> p;
        // initialize the helper vector used with all false
        used = vector<bool>(nums.size(), false);
        // start recursion
        genPerm(nums, 0, p);
        // return result
        return res;
    }
};

// test function
int main() {
    vector<int> nums = vector < int > {1, 2, 3, 4, 7};
    vector <vector<int>> res = Solution().permute(nums);
    for (int i = 0; i < res.size(); ++i) {
        for (int j = 0; j < res[i].size(); ++j) {
            cout << res[i][j] << " ";
        }
        cout << endl;
    }
    return 0;
}